pycontrails/models/apcemm/adapter.py

```diff
@@ -0,0 +1,811 @@
+    source: Flight
+
+    #: Output from trajectory calculation
+    trajectories: Flight | None
+
+    #: APCEMM time series output, starting from early vortex stage
+    vortex: pd.DataFrame | None
+
+    #: Paths to APCEMM netCDF output at prescribed time intervals
+    contrail: pd.DataFrame | None
+
+    #: Downsampling factor from trajectory time resolution to
+    #: APCEMM met input file resolution
+    _trajectory_downsampling: int
+
+    def __init__(
+        self,
+        met: MetDataset,
+        apcemm: str,
+        apcemm_root: str | None = None,
+        yaml: APCEMMYaml | None = None,
+        cachestore: cache.CacheStore | None = None,
+        params: dict[str, Any] | None = None,
+        **params_kwargs: Any,
+    ) -> None:
+        super().__init__(met, params=params, **params_kwargs)
+
+        self.apcemm = apcemm
+
+        if cachestore is None:
+            cache_root = cache._get_user_cache_dir()
+            cache_dir = f"{cache_root}/apcemm"
+            cachestore = cache.DiskCacheStore(cache_dir=cache_dir)
+        self.cachestore = cachestore
+
+        self._trajectory_downsampling = self._validate_downsampling()
+
+        if yaml is not None:
+            self.yaml = yaml
+        elif apcemm_root is not None:
+            self.yaml = APCEMMYaml(apcemm_root=apcemm_root)
+        else:
+            self.yaml = APCEMMYaml()
+
+    @overload
+    def eval(self, source: Flight, **params: Any) -> Flight: ...
+
+    @overload
+    def eval(self, source: None = ..., **params: Any) -> NoReturn: ...
+
+    def eval(self, source: Flight | None = None, **params: Any) -> Flight | NoReturn:
+        """Run APCEMM simulation on flight.
+
+        Simulates the formation and evolution of contrails from a Flight
+        using the APCEMM plume model described in Fritz et. al. (2020)
+        :cite:`schumannRolePlumescaleProcesses2020`.
+
+        Parameters
+        ----------
+        source : Flight | None
+            Input Flight to model.
+        **params : Any
+            Overwrite model parameters before eval.
+
+        Returns
+        -------
+        Flight | NoReturn
+            Flight with exit status of APCEMM simulations. Detailed APCEMM outputs are attached
+            to model :attr:`vortex` and :attr:`contrail` attributes (see :class:`APCEMM` notes
+            for details).
+
+        References
+        ----------
+        - :cite:`fritzRolePlumescaleProcesses2020`
+        """
+
+        self.update_params(params)
+        self.set_source(source)
+        self.source = self.require_source_type(Flight)
+
+        logger.info("Attaching APCEMM initial conditions to source")
+        self.attach_apcemm_initial_conditions()
+
+        logger.info("Computing Lagrangian trajectories")
+        self.compute_lagrangian_trajectories()
+
+        # Select segments to run in APCEMM
+        # Defaults to all segments, but allows user to select a subset
+        segments = self.params["segments"]
```

Comment:
These are segments. Make it clear and consistent which type of numerical simplification you use.

Reply:
It's definitely simplest to initialize APCEMM simulations at waypoints, not segments, so that's what the interface does now. Hopefully that comes across clearly in the code!

1dc2173b5f20445833c83ac16d15a3571d3001d9